File: packages/plugin-sdk/src/sparse-workerd/ensure.ts
/**
 * Downloads / refreshes the pinned Cloudflare `workerd` binary (mirrors `ensure.rs`).
 */

import { createHash } from "node:crypto";
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import { gunzipSync } from "node:zlib";
import path from "node:path";
import { fileURLToPath } from "node:url";
import os from "node:os";

/**
 * One platform asset entry inside {@link WorkerdPin}.
 */
export type WorkerdAsset = {
  /** Gzip-compressed release artifact filename on the Cloudflare workerd GitHub release. */
  artifact: string;
  /** Expected SHA-256 hex digest of the compressed artifact bytes (not the gunzipped binary). */
  sha256_hex: string;
};

/**
 * Contents of `workerd-pin.json` shipping with this package.
 */
export type WorkerdPin = {
  /** GitHub release tag (for example `v1.20250310.0`) that identifies this pin. */
  release_tag: string;
  /** Compatibility date bundled with this pin for Cap'n Proto configs. */
  bundled_compat_date: string;
  /** Stamp filename written beside the installed binary to record `release_tag`. */
  version_stamp: string;
  /** Platform key → artifact map (`linux-x86_64`, `macos-aarch64`, `windows-x86_64`, …). */
  assets: Record<string, WorkerdAsset>;
};

/**
 * Resolves the `@bookclerk/plugin-sdk` package root.
 *
 * Works whether this module is loaded from `src/sparse-workerd` or
 * `dist/sparse-workerd`.
 *
 * @returns Absolute path to `packages/plugin-sdk`.
 */
export function packageRoot(): string {
  const here = path.dirname(fileURLToPath(import.meta.url));
  // dist/sparse-workerd or src/sparse-workerd → package root
  return path.resolve(here, "../..");
}

/**
 * Loads and parses `workerd-pin.json` from the package root.
 *
 * @param root - Package root override (defaults to {@link packageRoot}).
 * @returns Parsed pin document.
 */
export function loadPin(root = packageRoot()): WorkerdPin {
  const pinPath = assertPathInside(path.resolve(root), "workerd-pin.json");
  return JSON.parse(fs.readFileSync(pinPath, "utf8")) as WorkerdPin;
}

/**
 * Maps Node `process.platform` / `process.arch` to a pin asset key.
 *
 * @param platform - Node platform string (default `process.platform`).
 * @param arch - Node arch string (default `process.arch`).
 * @returns Pin asset key, or `undefined` when the host is unsupported.
 */
export function platformKey(
  platform = process.platform,
  arch = process.arch,
): string | undefined {
  const map: Record<string, string> = {
    "linux-x64": "linux-x86_64",
    "linux-arm64": "linux-aarch64",
    "darwin-x64": "macos-x86_64",
    "darwin-arm64": "macos-aarch64",
    "win32-x64": "windows-x86_64",
  };
  return map[`${platform}-${arch}`];
}

/**
 * Returns the on-disk workerd binary basename for the current platform.
 *
 * @returns `workerd.exe` on Windows, otherwise `workerd`.
 */
export function binaryName(): string {
  return process.platform === "win32" ? "workerd.exe" : "workerd";
}

/**
 * Builds the GitHub release download URL for a pin artifact.
 *
 * @param pin - Loaded workerd pin.
 * @param artifact - Artifact filename from {@link WorkerdAsset.artifact}.
 * @returns Absolute HTTPS download URL.
 */
export function downloadUrl(pin: WorkerdPin, artifact: string): string {
  return `https://github.com/cloudflare/workerd/releases/download/${pin.release_tag}/${artifact}`;
}

/**
 * Resolve the workerd cache directory.
 *
 * Honors `BOOKCLERK_WORKERD_CACHE` as the selected install root (workspace
 * `target/`, `/opt`, external volumes). Otherwise uses
 * `~/.cache/bookclerk/workerd`. Derived files stay under that root.
 *
 * @returns Absolute cache directory path.
 */
export function defaultCacheDir(): string {
  const raw = process.env.BOOKCLERK_WORKERD_CACHE;
  if (raw && !raw.includes("\0")) {
    return path.resolve(raw);
  }
  return path.join(path.resolve(os.homedir()), ".cache", "bookclerk", "workerd");
}

/**
 * Resolves `candidate` under `root` via `path.resolve` + `path.relative` barrier.
 *
 * Rejects NUL bytes and `..` path components (names like `..draft` and
 * `edition..2` are allowed). Uses the CodeQL-recognized
 * `path.relative` / `startsWith("..")` containment pattern — no filesystem
 * probes before the barrier (those are sinks under local threat modeling).
 *
 * @param root - Trusted directory (resolved).
 * @param candidate - Absolute path or path relative to `root`.
 * @returns Absolute resolved path under `root`.
 * @throws {Error} When the path is empty, contains NUL/`..`, or escapes `root`.
 */
export function assertPathInside(root: string, candidate: string): string {
  if (!root || root.includes("\0") || !candidate || candidate.includes("\0")) {
    throw new Error("path is empty or contains NUL");
  }
  const normalized = candidate.replace(/\\/g, "/");
  if (normalized.split("/").includes("..")) {
    throw new Error(`path must not contain '..': ${candidate}`);
  }
  const resolvedRoot = path.resolve(root);
  const resolved = path.isAbsolute(candidate)
    ? path.resolve(candidate)
    : path.resolve(resolvedRoot, candidate);
  // CodeQL RelativePathStartsWithSanitizer / StartsWithDirSanitizer shape.
  const rel = path.relative(resolvedRoot, resolved);
  if (rel.startsWith(".." + path.sep) || rel === ".." || path.isAbsolute(rel)) {
    throw new Error(`path ${resolved} escapes root ${resolvedRoot}`);
  }
  if (!resolved.startsWith(resolvedRoot + path.sep) && resolved !== resolvedRoot) {
    throw new Error(`path ${resolved} escapes root ${resolvedRoot}`);
  }
  return resolved;
}

/**
 * Create `root` / `rel` as a directory after resolve + relative/`startsWith` barrier.
 *
 * @param root - Trusted directory.
 * @param rel - Relative suffix (may be multi-segment when each segment is safe).
 * @returns Absolute directory path under `root`.
 */
export function ensureDirUnder(root: string, rel: string): string {
  const resolvedRoot = path.resolve(root);
  const resolved = assertPathInside(resolvedRoot, rel);
  const relCheck = path.relative(resolvedRoot, resolved);
  if (
    relCheck.startsWith(".." + path.sep) ||
    relCheck === ".." ||
    path.isAbsolute(relCheck)
  ) {
    throw new Error(`path ${resolved} escapes root ${resolvedRoot}`);
  }
  // Inspect existing components before recursive mkdir so a symlink prefix
  // (e.g. `root/@bookclerk -> /outside`) cannot create `/outside/plugin-sdk`.
  refuseSymlinkExistingComponents(resolvedRoot, resolved);
  fs.mkdirSync(resolved, { recursive: true });
  return resolved;
}

/**
 * Write `contents` to `root` / `name` (single component) after resolve + barrier.
 *
 * Assumes `root` already exists (session/cache dirs are created first). No
 * mkdir of the tainted root — only write the barriered child path.
 *
 * @param root - Trusted directory (must already exist).
 * @param name - Single path component filename.
 * @param contents - Bytes or string to write.
 * @returns Absolute file path under `root`.
 */
export function writeFileUnder(
  root: string,
  name: string,
  contents: string | NodeJS.ArrayBufferView,
): string {
  singleFileComponent(name);
  const resolvedRoot = path.resolve(root);
  const resolved = path.resolve(resolvedRoot, name);
  const rel = path.relative(resolvedRoot, resolved);
  if (rel.startsWith(".." + path.sep) || rel === ".." || path.isAbsolute(rel)) {
    throw new Error(`path ${resolved} escapes root ${resolvedRoot}`);
  }
  if (!resolved.startsWith(resolvedRoot + path.sep) && resolved !== resolvedRoot) {
    throw new Error(`path ${resolved} escapes root ${resolvedRoot}`);
  }
  refuseSymlinkPath(resolvedRoot, resolved);
  fs.writeFileSync(resolved, contents);
  return resolved;
}

/**
 * Copy `src` to `root` / `name` after resolve + barrier on the destination.
 *
 * @param root - Trusted destination directory (must already exist).
 * @param name - Single path component filename.
 * @param src - Absolute source file (already validated by the caller).
 * @returns Absolute destination path under `root`.
 */
export function copyFileUnder(root: string, name: string, src: string): string {
  singleFileComponent(name);
  const resolvedRoot = path.resolve(root);
  const resolved = path.resolve(resolvedRoot, name);
  const rel = path.relative(resolvedRoot, resolved);
  if (rel.startsWith(".." + path.sep) || rel === ".." || path.isAbsolute(rel)) {
    throw new Error(`path ${resolved} escapes root ${resolvedRoot}`);
  }
  if (!resolved.startsWith(resolvedRoot + path.sep) && resolved !== resolvedRoot) {
    throw new Error(`path ${resolved} escapes root ${resolvedRoot}`);
  }
  refuseSymlinkPath(resolvedRoot, resolved);
  fs.copyFileSync(src, resolved);
  return resolved;
}

/**
 * Require `candidate` under `trustedRoot` with no symlink among *existing*
 * suffix components.
 *
 * Unlike {@link refuseSymlinkPath}, any trailing missing components are
 * allowed (not only the final one). Use before `mkdirSync(recursive)` so a
 * symlink intermediate cannot redirect directory creation outside the root.
 *
 * @param trustedRoot - Original operator/plugin root.
 * @param candidate - Path previously produced by {@link assertPathInside}.
 * @returns The validated absolute path.
 * @throws {Error} When an existing suffix component is a symlink or escapes.
 */
export function refuseSymlinkExistingComponents(
  trustedRoot: string,
  candidate: string,
): string {
  const rootLex = path.resolve(trustedRoot);
  const target = path.resolve(candidate);
  const rel = path.relative(rootLex, target);
  if (
    path.isAbsolute(rel) ||
    rel === ".." ||
    rel.startsWith(".." + path.sep) ||
    rel.split(path.sep).includes("..")
  ) {
    throw new Error(`path ${target} escapes root ${rootLex}`);
  }
  const root = fs.realpathSync(rootLex);
  const parts = rel === "" ? [] : rel.split(path.sep);
  let cur = root;
  for (let i = 0; i < parts.length; i++) {
    cur = path.resolve(cur, parts[i]!);
    const stepRel = path.relative(root, cur);
    if (
      path.isAbsolute(stepRel) ||
      stepRel === ".." ||
      stepRel.startsWith(".." + path.sep)
    ) {
      throw new Error(`path ${cur} escapes root ${root}`);
    }
    if (!cur.startsWith(root + path.sep) && cur !== root) {
      throw new Error(`path ${cur} escapes root ${root}`);
    }
    let st: fs.Stats;
    try {
      st = fs.lstatSync(cur);
    } catch (err) {
      const code = (err as NodeJS.ErrnoException).code;
      if (code === "ENOENT") {
        // Missing from here on: safe to create under the last existing prefix.
        break;
      }
      throw err;
    }
    if (st.isSymbolicLink()) {
      throw new Error(`refusing symlink in path: ${cur}`);
    }
  }
  return target;
}

/**
 * Require `candidate` under `trustedRoot` with no symlink suffix components.
 *
 * Resolves the trusted root identity once (a symlinked operator/plugin root is
 * allowed). Only path components *below* that identity are refused when they
 * are symlinks.
 *
 * @param trustedRoot - Original operator/plugin root.
 * @param candidate - Path previously produced by {@link assertPathInside}.
 * @returns The validated absolute path.
 * @throws {Error} When a suffix component is a symlink or escapes `trustedRoot`.
 */
export function refuseSymlinkPath(trustedRoot: string, candidate: string): string {
  const rootLex = path.resolve(trustedRoot);
  const target = path.resolve(candidate);
  const rel = path.relative(rootLex, target);
  if (
    path.isAbsolute(rel) ||
    rel === ".." ||
    rel.startsWith(".." + path.sep) ||
    rel.split(path.sep).includes("..")
  ) {
    throw new Error(`path ${target} escapes root ${rootLex}`);
  }
  // Allow the operator-selected root itself to be a symlink; constrain children.
  const root = fs.realpathSync(rootLex);
  const parts = rel === "" ? [] : rel.split(path.sep);
  let cur = root;
  for (let i = 0; i < parts.length; i++) {
    cur = path.resolve(cur, parts[i]!);
    const stepRel = path.relative(root, cur);
    if (
      path.isAbsolute(stepRel) ||
      stepRel === ".." ||
      stepRel.startsWith(".." + path.sep)
    ) {
      throw new Error(`path ${cur} escapes root ${root}`);
    }
    if (!cur.startsWith(root + path.sep) && cur !== root) {
      throw new Error(`path ${cur} escapes root ${root}`);
    }
    let st: fs.Stats;
    try {
      st = fs.lstatSync(cur);
    } catch (err) {
      const code = (err as NodeJS.ErrnoException).code;
      if (code === "ENOENT" && i === parts.length - 1) {
        break;
      }
      throw err;
    }
    if (st.isSymbolicLink()) {
      throw new Error(`refusing symlink in path: ${cur}`);
    }
  }
  return target;
}

/**
 * Validates a URL before `fetch` / download.
 *
 * Allows `https:` anywhere, or `http:` only to loopback hosts.
 *
 * @param url - Absolute URL string.
 * @returns Canonical href safe to request.
 * @throws {Error} When the URL is invalid or uses a disallowed scheme/host.
 */
export function validateFetchUrl(url: string): string {
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    throw new Error(`invalid URL: ${url}`);
  }
  if (parsed.protocol === "https:") {
    return parsed.href;
  }
  if (parsed.protocol === "http:") {
    const host = parsed.hostname.toLowerCase();
    if (
      host === "127.0.0.1" ||
      host === "localhost" ||
      host === "::1" ||
      host === "[::1]"
    ) {
      return parsed.href;
    }
  }
  throw new Error(
    `refusing non-HTTPS (or non-loopback HTTP) URL: ${parsed.protocol}//${parsed.host}`,
  );
}

/**
 * Validates a workerd (or helper) binary path before spawn.
 *
 * Requires an absolute path with no NUL bytes. When `trustedRoot` is set,
 * resolves both paths and requires the binary to stay under that root.
 *
 * @param bin - Candidate executable path.
 * @param trustedRoot - Optional directory the binary must remain under.
 * @returns Absolute validated path.
 * @throws {Error} When the path is relative, contains NUL, or escapes `trustedRoot`.
 */
export function validateSpawnExecutable(
  bin: string,
  trustedRoot?: string,
): string {
  if (!bin || bin.includes("\0")) {
    throw new Error("spawn executable path is empty or contains NUL");
  }
  if (!path.isAbsolute(bin)) {
    throw new Error(`spawn executable must be absolute: ${bin}`);
  }
  if (trustedRoot) {
    return assertPathInside(trustedRoot, bin);
  }
  return path.resolve(bin);
}

function singleFileComponent(name: string): void {
  if (
    !name ||
    name.includes("\0") ||
    name.includes("/") ||
    name.includes("\\") ||
    name === "." ||
    name === ".."
  ) {
    throw new Error(`file name must be a single path component: ${name}`);
  }
}

function stampFileName(pin: WorkerdPin): string {
  const stamp = pin.version_stamp;
  if (
    !stamp ||
    stamp.includes("\0") ||
    stamp.includes("/") ||
    stamp.includes("\\") ||
    stamp === ".." ||
    stamp === "."
  ) {
    throw new Error(`invalid workerd version_stamp: ${stamp}`);
  }
  return stamp;
}

function usableRegularFile(bin: string): boolean {
  try {
    const st = fs.lstatSync(bin);
    return st.isFile() && !st.isSymbolicLink();
  } catch {
    return false;
  }
}

/**
 * True when `bin` is a usable file matching `pin`.
 *
 * A sibling version stamp is enough when the binary exists. A stamp without a
 * binary is not current. When no Bookclerk stamp matches, probe `--version`
 * on the absolute path with fixed argv (`shell: false`, no `PATH` lookup).
 *
 * @param bin - Absolute candidate executable.
 * @param pin - Loaded workerd pin.
 * @returns Whether the binary is present and matches the pin.
 */
export function binaryMatchesPin(bin: string, pin: WorkerdPin): boolean {
  if (!usableRegularFile(bin)) return false;
  try {
    const dir = path.resolve(path.dirname(bin));
    const stamp = assertPathInside(dir, stampFileName(pin));
    const stampSt = fs.lstatSync(stamp);
    if (
      !stampSt.isSymbolicLink() &&
      stampSt.isFile() &&
      fs.readFileSync(stamp, "utf8").trim() === pin.release_tag
    ) {
      return true;
    }
  } catch {
    // Missing or invalid stamp → probe the absolute binary.
  }
  const validated = validateSpawnExecutable(bin);
  const out = spawnSync(validated, ["--version"], { encoding: "utf8", shell: false });
  if (out.status !== 0) return false;
  const combined = `${out.stdout ?? ""}${out.stderr ?? ""}`;
  const pinBare = pin.release_tag.replace(/^v/, "");
  return combined.includes(pin.release_tag) || combined.includes(pinBare);
}

/**
 * Ensures `cacheDir/workerd` matches the pin, downloading if needed.
 *
 * Honors `BOOKCLERK_WORKERD_BIN` when that absolute binary exists and matches
 * the pin (stamp or `--version`). The selected cache directory is the trusted
 * install root; children stay under it.
 *
 * @param cacheDir - Directory that will hold the binary (default {@link defaultCacheDir}).
 * @param root - Package root for loading the pin (default {@link packageRoot}).
 * @returns Absolute path to a current `workerd` binary.
 * @throws {Error} When the platform has no pin, download fails, or the digest mismatches.
 */
export async function ensureWorkerd(
  cacheDir = defaultCacheDir(),
  root = packageRoot(),
): Promise<string> {
  const pin = loadPin(root);
  const override = process.env.BOOKCLERK_WORKERD_BIN;
  if (override && !override.includes("\0") && path.isAbsolute(override)) {
    try {
      if (binaryMatchesPin(override, pin)) {
        return validateSpawnExecutable(override);
      }
    } catch {
      // Unusable override → fall through to cache install.
    }
  }

  const absCache = path.resolve(cacheDir);
  fs.mkdirSync(absCache, { recursive: true });
  const dest = assertPathInside(absCache, binaryName());
  refuseSymlinkPath(absCache, dest);
  if (binaryMatchesPin(dest, pin)) {
    return validateSpawnExecutable(dest, absCache);
  }

  const key = platformKey();
  if (!key || !pin.assets[key]) {
    throw new Error(
      `no pinned workerd asset for ${process.platform}-${process.arch}`,
    );
  }
  const asset = pin.assets[key]!;
  const url = validateFetchUrl(downloadUrl(pin, asset.artifact));
  console.error(`bookclerk-plugin: fetching ${url}`);
  // HTTPS (or loopback HTTP) URL validated by validateFetchUrl above.
  const res = await fetch(url);
  if (!res.ok) {
    throw new Error(`GET ${url} returned ${res.status}`);
  }
  const compressed = Buffer.from(await res.arrayBuffer());
  const got = createHash("sha256").update(compressed).digest("hex");
  if (got !== asset.sha256_hex) {
    throw new Error(
      `workerd download sha256 mismatch: got ${got}, expected ${asset.sha256_hex}`,
    );
  }

  const tmpName = `.${binaryName()}.tmp-${process.pid}-${Date.now()}-${Math.random().toString(16).slice(2)}`;
  const tmp = assertPathInside(absCache, tmpName);
  let ownedTmp = false;
  try {
    const fd = fs.openSync(tmp, "wx");
    ownedTmp = true;
    try {
      const binary = gunzipSync(compressed);
      fs.writeFileSync(fd, binary);
      if (process.platform !== "win32") {
        fs.fchmodSync(fd, 0o755);
      }
    } finally {
      fs.closeSync(fd);
    }
    refuseSymlinkPath(absCache, dest);
    fs.renameSync(tmp, dest);
    ownedTmp = false;
  } catch (err) {
    if (ownedTmp) {
      try {
        fs.unlinkSync(tmp);
      } catch {
        /* ignore */
      }
    }
    throw err;
  }
  writeFileUnder(absCache, stampFileName(pin), `${pin.release_tag}\n`);
  console.error(
    `bookclerk-plugin: installed ${pin.release_tag} → ${dest}`,
  );
  return validateSpawnExecutable(dest, absCache);
}

File: packages/plugin-sdk-python/src/bookclerk_plugin_sdk/sparse_workerd/ensure.py
"""Download / refresh the pinned Cloudflare ``workerd`` binary (mirrors ensure.rs).

Resolves the platform asset from ``workerd-pin.json``, verifies sha256, and
caches under the operator-selected directory (``BOOKCLERK_WORKERD_CACHE`` or
``~/.cache/bookclerk/workerd``).
"""

from __future__ import annotations

import gzip
import hashlib
import os
import platform
import shutil
import urllib.request
from pathlib import Path
from typing import Any

from ..path_guard import resolve_under, write_file_under

_PKG = Path(__file__).resolve().parent.parent  # bookclerk_plugin_sdk/


def package_root() -> Path:
    """Return the ``bookclerk_plugin_sdk`` package directory.

    Returns:
        Absolute path containing ``workerd.py``, ``bridge/``, and the pin file.
    """
    return _PKG


def load_pin(root: Path | None = None) -> dict[str, Any]:
    """Load the pinned workerd release metadata.

    Args:
        root: Package root containing ``workerd-pin.json`` (defaults to this SDK).

    Returns:
        Parsed pin dictionary (release tag, assets, version stamp).

    Raises:
        FileNotFoundError: If the pin file is missing.
        json.JSONDecodeError: If the pin file is not valid JSON.
    """
    pin_path = resolve_under(root or package_root(), "workerd-pin.json")
    import json

    return json.loads(pin_path.read_text(encoding="utf-8"))


def platform_key(
    system: str | None = None,
    machine: str | None = None,
) -> str | None:
    """Map OS/arch to a pin asset key such as ``linux-x86_64``.

    Args:
        system: OS name override (defaults to ``platform.system()``).
        machine: CPU arch override (defaults to ``platform.machine()``).

    Returns:
        Pin key string, or ``None`` when the platform is unsupported.
    """
    system = (system or platform.system()).lower()
    machine = (machine or platform.machine()).lower()
    os_map = {"linux": "linux", "darwin": "macos", "windows": "windows"}
    arch_map = {
        "x86_64": "x86_64",
        "amd64": "x86_64",
        "aarch64": "aarch64",
        "arm64": "aarch64",
    }
    os_name = os_map.get(system)
    arch = arch_map.get(machine)
    if not os_name or not arch:
        return None
    return f"{os_name}-{arch}"


def binary_name() -> str:
    """Return the on-disk workerd binary filename for this OS.

    Returns:
        ``workerd.exe`` on Windows, otherwise ``workerd``.
    """
    return "workerd.exe" if platform.system().lower() == "windows" else "workerd"


def download_url(pin: dict[str, Any], artifact: str) -> str:
    """Build the GitHub release download URL for a workerd artifact.

    Args:
        pin: Loaded pin metadata containing ``release_tag``.
        artifact: Asset filename from the pin ``assets`` table.

    Returns:
        Absolute HTTPS URL to the Cloudflare workerd release asset.
    """
    tag = pin["release_tag"]
    return f"https://github.com/cloudflare/workerd/releases/download/{tag}/{artifact}"


def default_cache_dir() -> Path:
    """Resolve the workerd binary cache directory.

    Honors ``BOOKCLERK_WORKERD_CACHE`` as the selected install root (any
    writable location, including paths outside ``$HOME`` and this package).
    Otherwise uses ``~/.cache/bookclerk/workerd``. Derived files stay under
    that root.

    Returns:
        Path to the cache directory (may not exist yet).
    """
    env = os.environ.get("BOOKCLERK_WORKERD_CACHE")
    if env and "\0" not in env:
        return Path(os.path.abspath(os.path.expanduser(env)))
    home = os.path.expanduser("~")
    return Path(os.path.abspath(os.path.join(home, ".cache", "bookclerk", "workerd")))


def validate_fetch_url(url: str) -> str:
    """Validate a URL before ``urlopen`` / download (request-forgery guard).

    Allows ``https:`` anywhere, or ``http:`` only to loopback hosts.

    Args:
        url: Absolute URL string.

    Returns:
        The same URL when it passes scheme/host checks.

    Raises:
        ValueError: When the URL is invalid or uses a disallowed scheme/host.
    """
    from urllib.parse import urlparse

    parsed = urlparse(url)
    if parsed.scheme == "https" and parsed.netloc:
        return url
    if parsed.scheme == "http":
        host = (parsed.hostname or "").lower()
        if host in {"127.0.0.1", "localhost", "::1"}:
            return url
    raise ValueError(
        f"refusing non-HTTPS (or non-loopback HTTP) URL: {parsed.scheme}://{parsed.netloc}"
    )


def validate_spawn_executable(
    bin_path: Path | str,
    trusted_root: Path | str | None = None,
) -> Path:
    """Validate a workerd (or helper) binary path before ``subprocess`` spawn.

    Requires an absolute path with no NUL bytes. When ``trusted_root`` is set,
    requires the binary to stay under that root via :func:`resolve_under`.

    Args:
        bin_path: Candidate executable path.
        trusted_root: Optional directory the binary must remain under.

    Returns:
        Absolute validated :class:`~pathlib.Path`.

    Raises:
        ValueError: When the path is relative, contains NUL, or escapes
            ``trusted_root``.
    """
    path = Path(bin_path)
    s = os.fspath(path)
    if not s or "\0" in s:
        raise ValueError("spawn executable path is empty or contains NUL")
    if not path.is_absolute():
        raise ValueError(f"spawn executable must be absolute: {path}")
    if trusted_root is not None:
        return resolve_under(trusted_root, path)
    return Path(os.path.abspath(s))


def _stamp_file_name(pin: dict[str, Any]) -> str:
    stamp = pin.get("version_stamp")
    if not isinstance(stamp, str) or not stamp:
        raise ValueError("invalid workerd version_stamp")
    if (
        "\0" in stamp
        or "/" in stamp
        or "\\" in stamp
        or stamp in {".", ".."}
    ):
        raise ValueError(f"invalid workerd version_stamp: {stamp}")
    return stamp


def _is_current(bin_path: Path, pin: dict[str, Any]) -> bool:
    """True when ``bin_path`` is a file matching the pin.

    A sibling stamp is enough when the binary exists. A stamp without a binary
    is not current. When no Bookclerk stamp matches, probe ``--version`` on the
    absolute path with fixed argv (no ``PATH`` lookup).
    """
    if not bin_path.is_file():
        return False
    root_s = os.path.abspath(os.fspath(bin_path.parent))
    name = _stamp_file_name(pin)
    stamp_s = os.path.abspath(os.path.join(root_s, name))
    try:
        rel = os.path.relpath(stamp_s, root_s)
    except ValueError:
        rel = ".."
    stamp_ok = (
        rel != ".."
        and not rel.startswith(".." + os.sep)
        and not os.path.isabs(rel)
        and (stamp_s == root_s or stamp_s.startswith(root_s + os.sep))
    )
    if stamp_ok:
        try:
            with open(stamp_s, encoding="utf-8") as fh:
                if fh.read().strip() == pin["release_tag"]:
                    return True
        except OSError:
            pass
    return _version_probe_matches(bin_path, pin)


def _version_probe_matches(bin_path: Path, pin: dict[str, Any]) -> bool:
    import subprocess

    validated = validate_spawn_executable(bin_path)
    try:
        out = subprocess.run(
            [os.fspath(validated), "--version"],
            check=False,
            capture_output=True,
            text=True,
            shell=False,
        )
    except OSError:
        return False
    if out.returncode != 0:
        return False
    combined = (out.stdout or "") + (out.stderr or "")
    tag = str(pin["release_tag"])
    bare = tag[1:] if tag.startswith("v") else tag
    return tag in combined or bare in combined


def ensure_workerd(
    cache_dir: Path | None = None,
    root: Path | None = None,
) -> Path:
    """Ensure a pinned ``workerd`` binary is available locally.

    Reuses ``BOOKCLERK_WORKERD_BIN`` or a cached binary when that file exists
    and the version stamp or ``--version`` output matches the pin. The selected
    cache directory is the trusted install root (not this package directory).
    A stamp without a binary is a cache miss.

    Args:
        cache_dir: Override cache directory (defaults to :func:`default_cache_dir`).
        root: SDK root for loading ``workerd-pin.json``.

    Returns:
        Path to an executable workerd binary matching the pin.

    Raises:
        RuntimeError: If no asset exists for this platform or the download hash
            mismatches.
        OSError: If the binary cannot be written or executed.
        ValueError: If a resolved binary path fails spawn validation.
    """
    pin = load_pin(root)

    override = os.environ.get("BOOKCLERK_WORKERD_BIN")
    if override and "\0" not in override:
        override_path = Path(override)
        if (
            override_path.is_absolute()
            and override_path.is_file()
            and _is_current(override_path, pin)
        ):
            return validate_spawn_executable(override_path)

    if cache_dir is None:
        cache_dir = default_cache_dir()
    cache_s = os.path.abspath(os.path.expanduser(os.fspath(cache_dir)))
    if "\0" in cache_s:
        raise ValueError("workerd cache path contains NUL")
    os.makedirs(cache_s, exist_ok=True)
    cache = Path(cache_s)

    dest = resolve_under(cache, binary_name())
    if dest.is_file() and _is_current(dest, pin):
        return validate_spawn_executable(dest, cache)

    key = platform_key()
    assets = pin.get("assets") or {}
    if not key or key not in assets:
        raise RuntimeError(
            f"no pinned workerd asset for {platform.system()}-{platform.machine()}"
        )
    asset = assets[key]
    url = validate_fetch_url(download_url(pin, asset["artifact"]))
    print(f"bookclerk-plugin: fetching {url}", flush=True)
    # HTTPS (or loopback HTTP) URL validated above.
    with urllib.request.urlopen(url) as resp:  # noqa: S310 — pinned GitHub release URL
        compressed = resp.read()
    got = hashlib.sha256(compressed).hexdigest()
    if got != asset["sha256_hex"]:
        raise RuntimeError(
            f"workerd download sha256 mismatch: got {got}, expected {asset['sha256_hex']}"
        )

    tmp_name = f"{binary_name()}.tmp"
    if (
        "/" in tmp_name
        or "\\" in tmp_name
        or ".." in tmp_name
        or tmp_name in {".", ".."}
    ):
        raise ValueError(f"invalid temp binary name: {tmp_name}")
    tmp_s = os.path.abspath(os.path.join(cache_s, tmp_name))
    try:
        rel = os.path.relpath(tmp_s, cache_s)
    except ValueError as err:
        raise ValueError(f"temp path escapes cache: {tmp_s}") from err
    if rel == ".." or rel.startswith(".." + os.sep) or os.path.isabs(rel):
        raise ValueError(f"temp path escapes cache: {tmp_s}")
    if tmp_s != cache_s and not tmp_s.startswith(cache_s + os.sep):
        raise ValueError(f"temp path escapes cache: {tmp_s}")
    with gzip.GzipFile(fileobj=__import__("io").BytesIO(compressed)) as gz, open(
        tmp_s, "wb"
    ) as out:
        shutil.copyfileobj(gz, out)
    if platform.system().lower() != "windows":
        os.chmod(tmp_s, 0o755)
    dest_s = os.fspath(dest)
    if dest_s != cache_s and not dest_s.startswith(cache_s + os.sep):
        raise ValueError(f"dest path escapes cache: {dest_s}")
    os.replace(tmp_s, dest_s)
    write_file_under(cache, _stamp_file_name(pin), f"{pin['release_tag']}\n")
    print(f"bookclerk-plugin: installed {pin['release_tag']} → {dest}", flush=True)
    return validate_spawn_executable(dest, cache)
